pkgs/os-specific/darwin/yabai/default.nix

```diff
@@ -20,6 +24,20 @@ stdenv.mkDerivation rec {
     SkyLight
   ];
 
+  buildPhase = ''
+    export PATH=/usr/bin:/bin:/usr/sbin
+
+    touch /tmp/.com.apple.dt.CommandLineTools.installondemand.in-progress;
+    PROD=$(softwareupdate -l |
+           grep "\*.*Command Line" |
+           head -n 1 | awk -F"*" '{print$2}' |
+           sed -e 's/^ *//' |
+           sed 's/Label: //g' |
+           tr -d '\n')
+    softwareupdate -i "$PROD" --verbose
+    /usr/bin/make install
+  '';
+
```

Comment:
You cannot force a software update in the user's macOS system. It's completely unreasonable that `nix-build` modifies system files.

Reply:
Yabai 4.0.1 requires Command Line Tools for `Xcode-*.*` to be installed. I tried to replace xcrun for crun but then the CI fails because if you call crun with a clean macos it prompts you to install xcode. So, what would be a more reasonable approach my good sir 🤔?

Reply:
I understand the issue, but if a package needs an up-to-date macOS SDK, I don't think you'll have any luck trying to build it. There is `darwin.apple_sdk_11_0.callPackage` that it is used by a few packages, but not newer.

But if your problem is related to the command line tools, you may be able to solve it by passing a different `stdenv` with:

```
      stdenv = if stdenv.cc.isClang then llvmPackages_XX.stdenv else stdenv;
```

Selecting a proper `llvmPackages_XX`.

There are a few examples, you have to change the attribute definition in `all-packages.nix`.

Reply:
Thanks for the throught explanation but I am not sure this is useful for yabai.
I do not think that this has to do with the apple sdk version TBH 😅, but I am not sure. If you look at the [travis.yml](https://github.com/koekeishiya/yabai/blob/master/.travis.yml) in yabai repo you will see that the image used is `osx_image: xcode10.1`.

Following the platform notes in [here](https://github.com/NixOS/nixpkgs/blob/80ffac409746c8b3d0696ee96cc25ccf7ac22a27/doc/stdenv/platform-notes.chapter.md). I replaced `xcrun crun` for `crun`.  If you check the diff comparing the makefile between the tags [4.0.1..3.3.10](
https://github.com/koekeishiya/yabai/compare/v3.3.10..v4.0.1#diff-beda42571c095172ab63437d050612a571d0d9ddd3ad4f2aecbce907a9b7e3d0R21-R28) . You will see that replacing `crun` for `xcrun crun` is part of the diff.

If I do this then when I build the derivation clang-11 complains about:
```bash
> clang-11: error: invalid arch name '-arch arm64e'
```
as you might be able to see in this [commit](https://github.com/NixOS/nixpkgs/pull/186108/commits/82e91a4951284936d81a80c50da2397ba157d1d1) [check run](https://github.com/NixOS/nixpkgs/pull/186108/checks?check_run_id=7843351557). 

So then I decided to replace this within [this commit](https://github.com/NixOS/nixpkgs/pull/186108/commits/8f94f9871498531b2f368a48f8ffd29991bc095f). Which thens produces this output and now I am stucked 😭 .

```bash
replacing old 'yabai-4.0.1'
installing 'yabai-4.0.1'
this derivation will be built:
  /nix/store/cyyjk14dgy5zzamwn61s2lfn5x3y32h0-yabai-4.0.1.drv
building '/nix/store/cyyjk14dgy5zzamwn61s2lfn5x3y32h0-yabai-4.0.1.drv'...
unpacking sources
unpacking source archive /nix/store/ny5hvzmdvmgfs0mxjrf1sph89bw4pdfg-source
source root is source
patching sources
configuring
no configure script, doing nothing
building
build flags: SHELL=/nix/store/db7n4xkfng8y4c3klp6bjppaws3rphq6-bash-5.1-p16/bin/bash
rm -rf ./bin
clang ./src/osax/loader.m -shared -O2 -mmacosx-version-min=10.13 -arch x86_64 -o ./src/osax/loader -framework Foundation
clang ./src/osax/payload.m -shared -fPIC -O2 -mmacosx-version-min=10.13 -arch x86_64  -o ./src/osax/payload -framework Foundation -framework Carbon
clang ./src/osax/mach_loader.m -O2 -mmacosx-version-min=10.13 -arch x86_64  -o ./src/osax/mach_loader -framework Cocoa
xxd -i -a ./src/osax/loader ./src/osax/loader_bin.c
xxd -i -a ./src/osax/payload ./src/osax/payload_bin.c
xxd -i -a ./src/osax/mach_loader ./src/osax/mach_loader_bin.c
rm -f ./src/osax/loader
rm -f ./src/osax/payload
rm -f ./src/osax/mach_loader
mkdir -p ./bin
clang src/manifest.m src/osax/loader_bin.c src/osax/payload_bin.c src/osax/mach_loader_bin.c -std=c99 -Wall -g -O0 -fvisibility=hidden -mmacosx-version-min=10.13 -fno-objc-arc -arch x86_64  -F/Sys
tem/Library/PrivateFrameworks -framework Carbon -framework Cocoa -framework CoreServices -framework SkyLight -framework ScriptingBridge -o bin/yabai
In file included from src/manifest.m:72:
src/workspace.m:109:31: error: property 'safeAreaInsets' not found on object of type 'NSScreen *'
                return screen.safeAreaInsets.top;
                              ^
1 error generated.
make: *** [makefile:59: bin/yabai] Error 1
error: builder for '/nix/store/cyyjk14dgy5zzamwn61s2lfn5x3y32h0-yabai-4.0.1.drv' failed with exit code 2;
       last 10 log lines:
       > rm -f ./src/osax/payload
       > rm -f ./src/osax/mach_loader
       > mkdir -p ./bin
       > clang src/manifest.m src/osax/loader_bin.c src/osax/payload_bin.c src/osax/mach_loader_bin.c -std=c99 -Wall -g -O0 -fvisibility=hidden -mmacosx-version-min=10.13 -fno-objc-arc -arch x86_6
4  -F/System/Library/PrivateFrameworks -framework Carbon -framework Cocoa -framework CoreServices -framework SkyLight -framework ScriptingBridge -o bin/yabai
       > In file included from src/manifest.m:72:
       > src/workspace.m:109:31: error: property 'safeAreaInsets' not found on object of type 'NSScreen *'
       >                 return screen.safeAreaInsets.top;
       >                               ^
       > 1 error generated.
       > make: *** [makefile:59: bin/yabai] Error 1
       For full logs, run 'nix log /nix/store/cyyjk14dgy5zzamwn61s2lfn5x3y32h0-yabai-4.0.1.drv'.

```
Maybe I am missing a framework? 🤔 , I am a bit lost TBH.

Reply:
It requires the macOS 12 SDK:

https://developer.apple.com/documentation/appkit/nsscreen/3882821-safeareainsets.

It won't build on current nixpkgs, where only these SDKs are available:

https://github.com/NixOS/nixpkgs/blob/af9e00071d0971eb292fd5abef334e66eda3cb69/pkgs/top-level/darwin-packages.nix#L35-L42

Reply:
Thanks for your reviews! 
I wonder how did you know that that docs are from the macOS12 SDK 😋. Do you reckon you could guide me into bringing this sdk into nixpgks? I think this one can be closed for the time being. What do you think?

Reply:
The documentation shows in which version is supported:
<img width="770" alt="Captura de Pantalla 2022-08-17 a la(s) 08 35 44" src="https://user-images.githubusercontent.com/238528/185134616-c6908bf4-d4af-459d-9a9b-1d3872db4c20.png">

I don't have enough knowledge to work in the macOS SDKs for nixpkgs, so I cannot comment about it. Only a few people can handle that work.